### PR TITLE
Add new article about Consul clients

### PR DESCRIPTION
**Pull request by ychuzevi**
Draft for article about Consul SDKs

### CODE REVIEW

File: consul/consul_client_03.md [hunk 1]
```diff
@@ -0,0 +1,28 @@
+# Be a good Consul client
+
+Consul servers may be distributed but their numbers is limited (we use cluster of 5 per datacenter). And if you are not careful, Consul requests may all be forwarded to the cluster's leader. On the other hand, you loose a lot of Consul's benefits if you restrict too much the calls to Consul.
```
Comment (pierresouchay) [inline consul/consul_client_03.md:5]: Add a very small intro describing what you will discuss in the article to give some context

File: consul/consul_client_03.md [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+
+At Criteo, we have around 3500 services registered in Consul, running in 12 datacenters for a total of 270 thousands instances. The behavior of Consul clients is hence really important in order to scale.
```
Comment (pierresouchay) [inline consul/consul_client_03.md]: It is 4k different services

File: consul/consul_client_03.md [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+
+## A brief history
+
+Most of our applications are C# or Java. When we migrated to Consul, we already had SDKs for those 2 languages. We relied on an in-house discovery system based on static registration in a database. And we provided full mesh between services thanks to client side load balancing libraries. The load balancing layer was taking care of doing the health checks and retrieving servers' weight in order to implement a weighted round robin algorithm.
```
Comment (pierresouchay) [inline consul/consul_client_03.md]: Trailing '

File: consul/consul_client_03.md [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+
+## Details that matters
+
+As explained in the previous article, the first point is that we rely heavily on [blocking queries](https://www.consul.io/api/features/blocking.html) (long polling). This allows to be notified immediately when a service changes and otherwise to keep the number of request really low.
+
+We also ensure that all our requests are made with [consistency parameter](https://www.consul.io/api/features/consistency.html) set to stale. This prevents the Consul servers to forward all the requests to the leader and ensure a better availability. To make sure that we are still serving fresh data, we are carefully monitoring the replication between the Consul servers. We actually went further and added an option to Consul [link = https://www.consul.io/docs/agent/options.html#discovery_max_stale].to also be able to change the default consistency behavior in order to have the same for projects not relying on SDKs (see [PR](https://github.com/hashicorp/consul/pull/3920)).
+
```
Comment (pierresouchay) [inline consul/consul_client_03.md:32]: You can also mention cache which allow reusing same request across several applications (iis/mesos)

File: consul/consul_client_03.md [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+Another useful configuration is that we enable [agent caching](https://www.consul.io/api/features/caching.html). This allows the local Consul agent to directly responds to a request without relying on the Consul servers if a similar request was made in the near past. This is especially useful for our applications that have multiple instances running on the same servers.
+
```
Comment (pierresouchay) [inline consul/consul_client_03.md:34]: We of course need a kind of conclusion.

**Review by pierresouchay — APPROVED** [review]
A typo, needs intro/conclusion, but very good start

File: consul/consul_client_03.md (hunk 1 above)
Comment (ychuzevi) [reply consul/consul_client_03.md:5]: Indeed will do

File: consul/consul_client_03.md (hunk 2 above)
Comment (ychuzevi) [reply consul/consul_client_03.md]: This number decreased a bit as I ran templaterb to check latest statistics but let's go for 4k for now in order to be consistent

File: consul/consul_client_03.md (hunk 4 above)
Comment (ychuzevi) [reply consul/consul_client_03.md:32]: See next paragraph but will try to develop a bit more.

File: consul/consul_client_03.md (hunk 5 above)
Comment (ychuzevi) [reply consul/consul_client_03.md:34]: Good point :)

**Review by pierresouchay — APPROVED** [review]

**Pull request merged**